File: src/integrationTest/java/net/ltgt/gradle/errorprone/StrongEncapsulationIntegrationTest.java
package net.ltgt.gradle.errorprone;

import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.TruthJUnit.assume;
import static java.util.Objects.requireNonNull;

import java.nio.file.Files;
import java.nio.file.StandardOpenOption;
import org.gradle.api.JavaVersion;
import org.gradle.testkit.runner.TaskOutcome;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class StrongEncapsulationIntegrationTest extends BasePluginIntegrationTest {
  private static final String FORKED = "%1$sFork: true%1$s".formatted(System.lineSeparator());
  private static final String NOT_FORKED = "%1$sFork: false%1$s".formatted(System.lineSeparator());
  private static final String JVM_ARG = "%sJVM Arg: ".formatted(System.lineSeparator());
  private static final String JVM_ARGS_STRONG_ENCAPSULATION =
      JVM_ARG + String.join(JVM_ARG, ErrorPronePlugin.JVM_ARGS_STRONG_ENCAPSULATION);

  private static String jvmArg(String argPrefix) {
    return JVM_ARG + argPrefix;
  }

  @BeforeEach
  void setup() throws Exception {
    Files.writeString(
        getBuildFile(),
        // language=kts
        """
        import net.ltgt.gradle.errorprone.*

        plugins {
            `java-library`
            id("%s")
        }
        repositories {
            mavenCentral()
        }
        dependencies {
            errorprone("com.google.errorprone:error_prone_core:%s")
        }

        val compileJava: JavaCompile by tasks
        val displayCompileJavaOptions by tasks.registering {
            doFirst {
                println("Fork: ${compileJava.options.isFork}")
                compileJava.options.forkOptions.allJvmArgs.forEach { arg ->
                    println("JVM Arg: $arg")
                }
            }
        }
        compileJava.finalizedBy(displayCompileJavaOptions)
        compileJava.options.forkOptions.jvmArgs!!.add("-XshowSettings")
        """
            .formatted(ErrorPronePlugin.PLUGIN_ID, errorproneVersion));
    writeSuccessSource();
  }

  @Test
  void doesNotConfigureForkingInJavaBefore16() throws Exception {
    assume()
        .withMessage("isJava16Compatible")
        .that(testJavaVersion)
        .isLessThan(JavaVersion.VERSION_16);

    // when
    var result = buildWithArgs("compileJava");
    // then
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
    assertThat(result.getOutput()).contains(NOT_FORKED);
    // Check that the configured jvm arg is preserved
    assertThat(result.getOutput()).contains(jvmArg("-XshowSettings"));

    // Test a forked task

    // given
    Files.writeString(
        getBuildFile(),
        // language=kts
        """

        compileJava.options.isFork = true
        """,
        StandardOpenOption.APPEND);

    // when
    result = buildWithArgs("compileJava");
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
    assertThat(result.getOutput()).contains(FORKED);
    assertThat(result.getOutput()).contains(JVM_ARGS_STRONG_ENCAPSULATION);
    // Check that the configured jvm arg is preserved
    assertThat(result.getOutput()).contains(jvmArg("-XshowSettings"));
  }

  @Test
  void configureForkingInJava16PlusVM() throws Exception {
    assume()
        .withMessage("isJava16Compatible")
        .that(testJavaVersion)
        .isAtLeast(JavaVersion.VERSION_16);

    // when
    var result = buildWithArgs("compileJava");
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
    assertThat(result.getOutput()).contains(FORKED);
    assertThat(result.getOutput()).contains(JVM_ARGS_STRONG_ENCAPSULATION);
    // Check that the configured jvm arg is preserved
    assertThat(result.getOutput()).contains(jvmArg("-XshowSettings"));

    // check that it doesn't mess with task avoidance

    // when
    result = buildWithArgs("compileJava");
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.UP_TO_DATE);
  }

  @Test
  void doesNotConfigureForkingIfErrorProneIsDisabled() throws Exception {
    assume()
        .withMessage("isJava16Compatible")
        .that(testJavaVersion)
        .isAtLeast(JavaVersion.VERSION_16);

    // given
    Files.writeString(
        getBuildFile(),
        // language=kts
        """

        compileJava.options.errorprone.enabled.set(false);
        """,
        StandardOpenOption.APPEND);

    // when
    var result = buildWithArgs("compileJava");

    // then
    assertThat(requireNonNull(result.task(":compileJava")).getOutcome())
        .isEqualTo(TaskOutcome.SUCCESS);
    assertThat(result.getOutput()).contains(NOT_FORKED);
    // Check that the configured jvm arg is preserved
    assertThat(result.getOutput()).contains(jvmArg("-XshowSettings"));
  }
}

File: src/integrationTest/kotlin/net/ltgt/gradle/errorprone/StrongEncapsulationIntegrationTest.kt
package net.ltgt.gradle.errorprone

import com.google.common.truth.Truth.assertThat
import com.google.common.truth.TruthJUnit.assume
import org.gradle.api.JavaVersion
import org.gradle.testkit.runner.TaskOutcome
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class StrongEncapsulationIntegrationTest : AbstractPluginIntegrationTest() {
    companion object {
        private val FORKED = "${System.lineSeparator()}Fork: true${System.lineSeparator()}"
        private val NOT_FORKED = "${System.lineSeparator()}Fork: false${System.lineSeparator()}"
        private val JVM_ARG = "${System.lineSeparator()}JVM Arg: "
        private val JVM_ARGS_STRONG_ENCAPSULATION =
            ErrorPronePlugin.JVM_ARGS_STRONG_ENCAPSULATION.joinToString(
                prefix = JVM_ARG,
                separator = JVM_ARG,
            )

        private fun jvmArg(argPrefix: String) = "$JVM_ARG$argPrefix"
    }

    @BeforeEach
    fun setup() {
        buildFile.appendText(
            """
            plugins {
                `java-library`
                id("${ErrorPronePlugin.PLUGIN_ID}")
            }
            repositories {
                mavenCentral()
            }
            dependencies {
                errorprone("com.google.errorprone:error_prone_core:$errorproneVersion")
            }

            val compileJava: JavaCompile by tasks
            val displayCompileJavaOptions by tasks.creating {
                doFirst {
                    println("Fork: ${'$'}{compileJava.options.isFork}")
                    compileJava.options.forkOptions.allJvmArgs.forEach { arg ->
                        println("JVM Arg: ${'$'}arg")
                    }
                }
            }
            compileJava.finalizedBy(displayCompileJavaOptions)
            compileJava.options.forkOptions.jvmArgs!!.add("-XshowSettings")
            """.trimIndent(),
        )
        testProjectDir.writeSuccessSource()
    }

    @Test
    fun `does not configure forking in Java before 16`() {
        assume().withMessage("isJava16Compatible").that(testJavaVersion).isLessThan(JavaVersion.VERSION_16)

        // when
        testProjectDir.buildWithArgs("compileJava").also { result ->
            // then
            assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
            assertThat(result.output).contains(NOT_FORKED)
            // Check that the configured jvm arg is preserved
            assertThat(result.output).contains(jvmArg("-XshowSettings"))
        }

        // Test a forked task

        // given
        buildFile.appendText(
            """

            compileJava.options.isFork = true
            """.trimIndent(),
        )

        // when
        testProjectDir.buildWithArgs("compileJava").also { result ->
            // then
            assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
            assertThat(result.output).contains(FORKED)
            assertThat(result.output).contains(JVM_ARGS_STRONG_ENCAPSULATION)
            // Check that the configured jvm arg is preserved
            assertThat(result.output).contains(jvmArg("-XshowSettings"))
        }
    }

    @Test
    fun `configure forking in Java 16+ VM`() {
        assume().withMessage("isJava16Compatible").that(testJavaVersion).isAtLeast(JavaVersion.VERSION_16)

        // when
        testProjectDir.buildWithArgs("compileJava").also { result ->
            // then
            assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
            assertThat(result.output).contains(FORKED)
            assertThat(result.output).contains(JVM_ARGS_STRONG_ENCAPSULATION)
            // Check that the configured jvm arg is preserved
            assertThat(result.output).contains(jvmArg("-XshowSettings"))
        }

        // check that it doesn't mess with task avoidance

        // when
        testProjectDir.buildWithArgs("compileJava").also { result ->
            // then
            assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.UP_TO_DATE)
        }
    }

    @Test
    fun `does not configure forking if Error Prone is disabled`() {
        assume().withMessage("isJava16Compatible").that(testJavaVersion).isAtLeast(JavaVersion.VERSION_16)

        // given
        buildFile.appendText(
            """

            compileJava.options.errorprone.enabled.set(false)
            """.trimIndent(),
        )

        // when
        val result = testProjectDir.buildWithArgs("compileJava")

        // then
        assertThat(result.task(":compileJava")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
        assertThat(result.output).contains(NOT_FORKED)
        // Check that the configured jvm arg is preserved
        assertThat(result.output).contains(jvmArg("-XshowSettings"))
    }
}
